File: sdks/go/hellomail/templates.go
package hellomail

import "context"

// TemplatesResource provides access to the template management API.
type TemplatesResource struct {
	client *Client
}

// Create creates a new email template.
func (r *TemplatesResource) Create(ctx context.Context, params *CreateTemplateParams) (*Template, error) {
	var tmpl Template
	if err := r.client.do(ctx, "POST", "/v1/templates", params, &tmpl); err != nil {
		return nil, err
	}
	return &tmpl, nil
}

// Get retrieves a single template by its ID.
func (r *TemplatesResource) Get(ctx context.Context, id string) (*Template, error) {
	var tmpl Template
	if err := r.client.do(ctx, "GET", "/v1/templates/"+id, nil, &tmpl); err != nil {
		return nil, err
	}
	return &tmpl, nil
}

// List returns all templates for the authenticated organization.
func (r *TemplatesResource) List(ctx context.Context) ([]Template, error) {
	var templates []Template
	if err := r.client.do(ctx, "GET", "/v1/templates", nil, &templates); err != nil {
		return nil, err
	}
	return templates, nil
}

// Update modifies an existing template.
func (r *TemplatesResource) Update(ctx context.Context, id string, params *UpdateTemplateParams) (*Template, error) {
	var tmpl Template
	if err := r.client.do(ctx, "PATCH", "/v1/templates/"+id, params, &tmpl); err != nil {
		return nil, err
	}
	return &tmpl, nil
}

// Delete removes a template.
func (r *TemplatesResource) Delete(ctx context.Context, id string) error {
	return r.client.do(ctx, "DELETE", "/v1/templates/"+id, nil, nil)
}

// Preview renders a template with the given data and returns the result.
func (r *TemplatesResource) Preview(ctx context.Context, id string, params *PreviewTemplateParams) (*TemplatePreview, error) {
	var preview TemplatePreview
	if err := r.client.do(ctx, "POST", "/v1/templates/"+id+"/preview", params, &preview); err != nil {
		return nil, err
	}
	return &preview, nil
}

File: sdks/node/src/resources/templates.ts
import type { Mailngine } from '../client.js';
import type {
  CreateTemplateParams,
  PreviewTemplateParams,
  Template,
  TemplatePreview,
  UpdateTemplateParams,
} from '../types.js';

/**
 * Resource for managing email templates.
 */
export class TemplatesResource {
  constructor(private readonly client: Mailngine) {}

  /**
   * Create a new template.
   *
   * POST /v1/templates
   */
  async create(params: CreateTemplateParams): Promise<Template> {
    return this.client.request<Template>('/v1/templates', {
      method: 'POST',
      body: params,
    });
  }

  /**
   * List all templates for the current organization.
   *
   * GET /v1/templates
   */
  async list(): Promise<Template[]> {
    return this.client.request<Template[]>('/v1/templates');
  }

  /**
   * Retrieve a single template by ID.
   *
   * GET /v1/templates/:id
   */
  async get(id: string): Promise<Template> {
    return this.client.request<Template>(`/v1/templates/${encodeURIComponent(id)}`);
  }

  /**
   * Update a template.
   *
   * PATCH /v1/templates/:id
   */
  async update(id: string, params: UpdateTemplateParams): Promise<Template> {
    return this.client.request<Template>(`/v1/templates/${encodeURIComponent(id)}`, {
      method: 'PATCH',
      body: params,
    });
  }

  /**
   * Delete a template.
   *
   * DELETE /v1/templates/:id
   */
  async delete(id: string): Promise<void> {
    await this.client.request<unknown>(`/v1/templates/${encodeURIComponent(id)}`, {
      method: 'DELETE',
    });
  }

  /**
   * Preview a template with sample data.
   *
   * POST /v1/templates/:id/preview
   */
  async preview(id: string, params: PreviewTemplateParams): Promise<TemplatePreview> {
    return this.client.request<TemplatePreview>(
      `/v1/templates/${encodeURIComponent(id)}/preview`,
      { method: 'POST', body: params },
    );
  }
}
